java/flight/flight-core/src/main/java/org/apache/arrow/flight/auth2/BasicAuthValidator.java

```diff
@@ -0,0 +1,147 @@
+  private final AuthTokenManager authTokenManager;
+
+  /**
+   * Creates a validator with supplied AuthTokenManager to generate token
+   * and CredentialValidator to validate username/password. 
+   * @param authTokenManager AuthTokenManager used to generate tokens.
+   * @param credentialValidator CredentialValidator used to validate the username/password
+   */
+  public BasicAuthValidator(CredentialValidator credentialValidator, AuthTokenManager authTokenManager) {
+    super();
+    this.credentialValidator = credentialValidator;
+    this.authTokenManager = authTokenManager;
+  }
+
+  /**
+   * Interface that this validator delegates for determining if the credentials are valid.
+   */
+  interface CredentialValidator {
+    Optional<String> validate(String username, String password) throws Exception;
+  }
+
+  /**
+   * Interface that this validator delegates for generating and validating a token.
+   */
+  interface AuthTokenManager {
+    Optional<String> generateToken(String username, String password);
+
+    Optional<String> validateToken(String token);
+  }
+
+  @Override
+  public AuthResult validateIncomingHeaders(CallHeaders incomingHeaders) throws Exception {
+    AuthResult authResult = parseAndValidateBearerHeaders(incomingHeaders);
+    if (authResult != null) {
+      return authResult;
+    }
+    return parseAndValidateBasicHeaders(incomingHeaders);
+  }
+
+  /**
+   * Appends an authorization header with a bearer token to the outgoing headers.
+   * @param outgoingHeaders Outgoing headers to append the authorization header to.
+   * @param token Token to be appended to the outgoing header along with the authorization header.
+   */
+  private void appendAuthHeaderWithBearerTokenToOutgoingHeaders(CallHeaders outgoingHeaders, String token) {
+    if (null == AuthUtilities.getValueFromAuthHeader(outgoingHeaders, Auth2Constants.BEARER_PREFIX)) {
+      outgoingHeaders.insert(Auth2Constants.AUTHORIZATION_HEADER,
+              Auth2Constants.BEARER_PREFIX + token);
+    }
+  }
+
+  /**
+   * Parses and validates basic headers and returns the AuthResult.
+   * @param incomingHeaders Incoming header to parse.
+   * @return An instance of an AuthResult.
+   * @throws Exception When the basic header does not contain the credentials or the credential validation fails.
+   */
+  private AuthResult parseAndValidateBasicHeaders(CallHeaders incomingHeaders) throws Exception {
+    final String authEncoded = AuthUtilities.getValueFromAuthHeader(incomingHeaders, Auth2Constants.BASIC_PREFIX);
+    if (authEncoded == null) {
+      throw CallStatus.UNAUTHENTICATED.toRuntimeException();
+    }
+    // The value has the format Base64(<username>:<password>)
+    final String authDecoded = new String(Base64.getDecoder().decode(authEncoded), StandardCharsets.UTF_8);
+    final int colonPos = authDecoded.indexOf(':');
+    if (colonPos == -1) {
+      throw CallStatus.UNAUTHORIZED.toRuntimeException();
```

Comment:
Should this be unauthenticated?

Reply:
Yes, this should be UNAUTHENTICATED, updated code. I believe yes, this should throw so that authentication fails and the exception is caught in the authenticate method.